expression/constant_fold.go

```diff
@@ -86,19 +86,36 @@ func foldConstant(expr Expression) (Expression, bool) {
 		}
 
 		args := x.GetArgs()
-		canFold := true
+		sc := x.GetCtx().GetSessionVars().StmtCtx
+		nonConstArgIdx := make([]int, 0, len(args))
+		hasNullArg := false
 		isDeferredConst := false
 		for i := 0; i < len(args); i++ {
 			foldedArg, isDeferred := foldConstant(args[i])
 			x.GetArgs()[i] = foldedArg
-			_, conOK := foldedArg.(*Constant)
-			if !conOK {
-				canFold = false
+			con, conOK := foldedArg.(*Constant)
+			if conOK {
+				if con.Value.IsNull() {
+					hasNullArg = true
+				}
+			} else {
+				nonConstArgIdx = append(nonConstArgIdx, i)
 			}
 			isDeferredConst = isDeferredConst || isDeferred
 		}
-		if !canFold {
-			return expr, isDeferredConst
+		if len(nonConstArgIdx) > 0 {
+			if !hasNullArg || !sc.InNullRejectCheck {
```

Comment:
Do we need to check whether it's in null-reject?

Reply:
Yes we need, for cases like `select null and a < 1 from t;`, if we fold null constant for the projection part, the result would be `null` for all rows, but actually for `t.a >= 1`, the result should be `0`